File: Sort-Algorithms-2.0/Python/Counting_Sort.py
import os
from time import time
import random

def CountingSort(arr): 
    max_element = int(max(arr)) 
    min_element = int(min(arr)) 
    range_of_elements = max_element - min_element + 1

    count_arr = [0 for _ in range(range_of_elements)] 
    output_arr = [0 for _ in range(len(arr))] 
  
    for i in range(0, len(arr)): 
        count_arr[arr[i]-min_element] += 1
  
    for i in range(1, len(count_arr)): 
        count_arr[i] += count_arr[i-1] 
  
    for i in range(len(arr)-1, -1, -1): 
        output_arr[count_arr[arr[i] - min_element] - 1] = arr[i] 
        count_arr[arr[i] - min_element] -= 1
  
    for i in range(0, len(arr)): 
        arr[i] = output_arr[i] 
  
    return arr 


def crear_arreglo(A,n):
    for i in range(n):
        A.append(random.randint(1,1000000))


if __name__=='__main__':
    A=[1000000]
    f=open("tiempos_counting_p.txt","w")
    for i in range(100000,1000001,100000):
        crear_arreglo(A,i)
        tiempo_inicial=time()
        CountingSort(A)
        tiempo_final=time()
        tiempo_ejecucion=tiempo_final-tiempo_inicial
        f.write(str(tiempo_ejecucion))
        print("tiempo de ejecucion: ",tiempo_ejecucion)
    f.close()

File: Sort-Algorithms-2.0/C++/Counting_Sort.cpp
#include <algorithm>
#include<iostream>
#include<time.h>
#include<fstream>
#include<stdlib.h>
#include <vector>
using namespace std;

void countSort(vector<int>& arr,int i)
{
    int max = *max_element(arr.begin(), arr.end());
    int min = *min_element(arr.begin(), arr.end());
    int range = max - min + 1;

    vector<int> count(range), output(arr.size());
    for (int i = 0; i < arr.size(); i++)
        count[arr[i] - min]++;

    for (int i = 1; i < count.size(); i++)
        count[i] += count[i - 1];

    for (int i = arr.size() - 1; i >= 0; i--) {
        output[count[arr[i] - min] - 1] = arr[i];
        count[arr[i] - min]--;
    }

    for (int i = 0; i < arr.size(); i++)
        arr[i] = output[i];
}

void printArray(vector<int>& arr)
{
    for (int i = 0; i < arr.size(); i++)
        cout << arr[i] << " ";
    cout << "\n";
}


void crear_arreglo(vector<int>& v,int n){
    int c;
    srand(time(NULL));

    for(c=0;c<=n;c++){
        v[c]=rand()%1000000;

    }
}


int main()
{
    unsigned t0,t1;
    double time;
    vector<int> v(1000000);
    ofstream archivo;
    archivo.open("tiempos_counting_c.txt",ios::out);
    for(int i=100000;i<=1000000;i+=100000){
        crear_arreglo(v,i);
        t0=clock();
        countSort(v,i-1);
        t1=clock();
        time=(double(t1-t0)/CLOCKS_PER_SEC);
        cout<<"Tiempo de ejecucion: "<<time<<endl;
        archivo<<time<<endl;
    }
    archivo.close();
    return 0;
}
